File: libs/power-log-parser/src/lib/parsers/game-cleanup-parser.ts
import { GameTag } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider } from '../game-event';
import { Node, NodeType, TagChange } from '../models';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

const STATE_COMPLETE = 3;

export class GameCleanupParser implements ActionParser {
	readonly ParserName = 'GameCleanupParser';

	private ParserState: ParserState;

	constructor(parserState: ParserState, _stateFacade: StateFacade) {
		this.ParserState = parserState;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.GameState &&
			node.Type === NodeType.TagChange &&
			(node.Object as TagChange).Name === (GameTag.STATE as number) &&
			(node.Object as TagChange).Value === STATE_COMPLETE
		);
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		console.debug('Ending current game in GS', node.CreationLogLine);
		this.ParserState.EndCurrentGame();
		return null;
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/GameCleanupParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;

namespace HearthstoneReplays.Events.Parsers
{
    public class GameCleanupParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public GameCleanupParser(ParserState ParserState, StateFacade helper)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = helper;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            // Simply to be able to mark the game as ended
            return stateType == StateType.GameState
                && node.Type == typeof(TagChange)
                && (((node.Object as TagChange).Name == (int)GameTag.STATE
                                && (node.Object as TagChange).Value == (int)State.COMPLETE));
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            Logger.Log("Ending current game in GS", node.CreationLogLine);
            ParserState.EndCurrentGame();
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
